main/src/ca/uwaterloo/flix/language/errors/ParseError.scala

```diff
@@ -18,28 +18,64 @@ package ca.uwaterloo.flix.language.errors
 
 import ca.uwaterloo.flix.language.CompilationMessage
 import ca.uwaterloo.flix.language.ast.Ast.SyntacticContext
-import ca.uwaterloo.flix.language.ast.SourceLocation
+import ca.uwaterloo.flix.language.ast.{SourceLocation, TokenKind}
 import ca.uwaterloo.flix.util.Formatter
 
 /**
- * An error raised to indicate a parse error.
- *
- * @param msg the error message.
- * @param ctx the syntactic context.
- * @param loc the source location.
- */
-case class ParseError(msg: String, ctx: SyntacticContext, loc: SourceLocation) extends CompilationMessage with Recoverable {
-  val kind = "Parse Error"
+  * An error raised to indicate an unexpected token was found.
+  *
+  * @param expected Names of the tokens that are expected at the location. See [[TokenKind.display]].
+  * @param actual   Name of the token that was actually found. See [[TokenKind.display]]
+  * @param ctx      The syntactic context.
+  * @param loc      The source location.
+  * @param hint     Optional hint with more details about the error
+  */
+case class UnexpectedToken(expected: Seq[String], actual: String, ctx: SyntacticContext, loc: SourceLocation, hint: Option[String] = None) extends CompilationMessage with Recoverable {
+  val kind = s"Parse Error ($ctx)"
+
+  /**
+    * Joins items nicely with comma separation ending with an "or".
+    * For instance prettyJoin(List("def", "enum", "trait")) gives "def, enum or trait".
+    */
+  private def prettyJoin[T](items: Seq[T]): String = items match {
+    case i1 :: i2 :: Nil => s"$i1 or $i2"
+    case i1 :: Nil => s"$i1"
+    case i :: tail => s"$i, ${prettyJoin(tail)}"
+  }
 
-  def summary: String = msg
+  def summary: String = s"Expected ${prettyJoin(expected)} before $actual"
 
   def message(formatter: Formatter): String = {
     import formatter._
+    val hintStr = hint.map(s"\nHint: " + _).getOrElse("")
+    val expectedStr = prettyJoin(expected.map(cyan))
     s"""${line(kind, source.name)}
-       |>> Parse Error: ${red(msg)}
+       |>> Expected $expectedStr before ${red(actual)}
+       |
+       |${code(loc, s"Here")}$hintStr
+       |""".stripMargin
+  }
+}
+
+/**
+  * An generic parse error. We should prefer [[UnexpectedToken]] when possible.
+  *
+  * @param message Error message to be displayed.
+  * @param ctx     The syntactic context.
+  * @param loc     The source location.
+  * @param hint    Optional hint with more details about the error
+  */
+case class ParseError(message: String, ctx: SyntacticContext, loc: SourceLocation, hint: Option[String] = None) extends CompilationMessage with Recoverable {
```

Comment:
The "message" here is an anti-pattern. It means that we cannot use the type to determine which specific error is in play.

Reply:
I see, I'll do another revision then.

Reply:
It helps structure the code. Consider WeederError, TypeError etc.

All could have a `message: String`, but it provides no structure.

Reply:
Yep, I completely agree. I made something much more akin to `WeederError` now. Could you have another look?